### PR TITLE
Kanban Set 2 Review 2 (dev->staging)

### PR DESCRIPTION
**Pull request by afrasier**
Includes:

- #270 
- #274 
- #275 
- #277 
- #279 
- #281 
- #282 
- #287 
- #288

### CODE REVIEW

File: talentmap_api/common/common_helpers.py [hunk 1]
```diff
@@ -24,6 +26,31 @@ def resolve_path_to_view(request_path):
     return view
 
 
+def safe_navigation(object, attribute):
```
Comment (jseppi) [inline talentmap_api/common/common_helpers.py]: 👍

File: talentmap_api/common/common_helpers.py [hunk 2]
```diff
@@ -42,6 +69,24 @@ def get_prefetched_filtered_queryset(model, serializer_class, *args, **kwargs):
     return queryset
 
 
+def ensure_date(date):
```
Comment (jseppi) [inline talentmap_api/common/common_helpers.py]: 👍

File: talentmap_api/language/models.py [hunk 3]
```diff
@@ -49,6 +48,18 @@ class Proficiency(StaticRepresentationModel):
     def __str__(self):
         return f"{self.code}"
 
+    def __gt__(self, other):
```
Comment (jseppi) [inline talentmap_api/language/models.py]: nice

File: talentmap_api/user_profile/models.py [hunk 4]
```diff
@@ -57,6 +52,48 @@ def is_cdo(self):
 
         return self.direct_reports.count() != 0
 
+    @property
+    def is_six_eight(self):
+        '''
+        Determines if this user is classified as a 6/8 valid bidder. The rules to calculate this have some
+        exceptions, but this is just a baseline logic implmentation for the time being.
```
Comment (jseppi) [inline talentmap_api/user_profile/models.py]: `implmentation` -> `implementation`

File: talentmap_api/user_profile/models.py [hunk 5]
```diff
@@ -57,6 +52,48 @@ def is_cdo(self):
...
+        '''
+        # An employee is valid under 6/8 rules if the following states are TRUE
+        #  - Has served a suffiecient portion of a non-domestic tour of duty (10 of 12, 20 of 24, 30 of 36 or 83% of others)
```
Comment (jseppi) [inline talentmap_api/user_profile/models.py]: `suffiecient` -> `sufficient`

File: talentmap_api/user_profile/models.py [hunk 6]
```diff
@@ -57,6 +52,48 @@ def is_cdo(self):
...
+        # Count our assignments in order from most recent to oldest, counting the service duration until we
+        # hit a foreign assignment, which is now guaranteed to be a valid duration to break apart 6/8 tabulations
+        total = 0
+        for assignment in list(assignments):
+            if assignment.position.post.location.country.code != "USA":
+                break
+            total += assignment.service_duration
+
+        # If the total number of now contiguous domestic service exceeds 6 years we should return FALSE as we don't meet the requirements
+        return total < (6 * 12)
```
Comment (jseppi) [inline talentmap_api/user_profile/models.py]: This is some funky logic (not your code, but the 6/8 calculation itself)!

File: talentmap_api/bidding/serializers/prepanel.py [hunk 7]
```diff
@@ -0,0 +1,90 @@
...
+    def generate_prepanel_six_eight(self, user, position, waivers, sii):
+        return {
+            "calculated": user.is_six_eight,
+            "self_identified": sii.is_six_eight,
+            "post_location": str(safe_navigation(position, 'post.location')),
+            "waivers": [str(x) for x in list(waivers.filter(category=Waiver.Category.six_eight))]
+        }
+
+    def generate_prepanel_language(self, user, position, waivers):
+        language_match = True
```
[the remark this answers is not in the document]
Comment (afrasier) [reply talentmap_api/bidding/serializers/prepanel.py]: Yup, will add comment

File: talentmap_api/bidding/serializers/prepanel.py [hunk 8]
```diff
@@ -0,0 +1,90 @@
...
+            "user_languages": [str(x) for x in list(user.language_qualifications.all())],
+            "waviers": [str(x) for x in list(waivers.filter(category=Waiver.Category.language))]
+        }
+
+    def generate_prepanel_skill(self, user, position, waivers):
+        return {
+            "skill_match": user.skill_code.filter(code=safe_navigation(position, 'skill.code')).exists(),
+            "position_skill": str(safe_navigation(position, 'skill')),
+            "user_skills": [str(x) for x in list(user.skill_code.all())],
+            "waviers": [str(x) for x in list(waivers.filter(category=Waiver.Category.skill))]
```
[the remark this answers is not in the document]
Comment (afrasier) [reply talentmap_api/bidding/serializers/prepanel.py]: Easily addable later, didn't want to make the prepanel object too big before we get some testing in.

File: talentmap_api/bidding/serializers/serializers.py [hunk 9]
```diff
@@ -65,6 +111,14 @@ class BidSerializer(PrefetchedSerializer):
     position = StaticRepresentationField(read_only=True)
     waivers = StaticRepresentationField(read_only=True, many=True)
 
+    is_paneling_today = serializers.SerializerMethodField()
+
+    def get_is_paneling_today(self, obj):
+        '''
+        We provide this so that if the front end is accessed from a different time zone, we are consistent
```
[the remark this answers is not in the document]
Comment (afrasier) [reply talentmap_api/bidding/serializers/serializers.py]: Yes, I suppose that comment is a bit confusing. I wrote that before I made it a property on the bid; I'll actually just refactor this.

File: talentmap_api/saml2/attribute_maps/basic.py [hunk 10]
```diff
@@ -0,0 +1,325 @@
+MAP = {
```
[the remark this answers is not in the document]
Comment (afrasier) [reply talentmap_api/saml2/attribute_maps/basic.py]: 🤞

File: talentmap_api/user_profile/models.py (hunk 6 above)
Comment (afrasier) [reply talentmap_api/user_profile/models.py]: Haha, it's worse when you start seeing all the exceptions.

Comment (jseppi) [reply talentmap_api/user_profile/models.py]: 😬 💀
Well, this should be a HUGE improvement for the users that were previously doing this by hand/excel/sorting hat.

**Pull request merged**